### PR TITLE
Fix oss-fuzz build compilation errors

### PR DESCRIPTION
**Pull request by disconnect3d**
This PR fixes two compilation error that are happening in the oss-fuzz build of Envoy. Note that the latest oss-fuzz build errors out with only the [ error related to `AsyncStream`](https://oss-fuzz-build-logs.storage.googleapis.com/log-917e4a97-e528-4cce-b749-9187973dd9ed.txt) but AFAIR when it is fixed, a different compilation error has to be fixed which I fixed in the first commit that changes the socket.h file.

### CODE REVIEW

File: envoy/network/socket.h [hunk 1]
```diff
@@ -23,7 +23,6 @@ namespace Network {
 // avoid #ifdef proliferation.
 struct SocketOptionName {
   SocketOptionName() = default;
-  SocketOptionName(const SocketOptionName&) = default;
   SocketOptionName& operator=(const SocketOptionName&) = default;
```
Comment (yanavlasov) [inline envoy/network/socket.h:26]: You can also remove declaration of the default assignment operator. Same below.

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

**Comment by yanavlasov** [comment]
Please also fix DCO. You can click on the "Details" link next to it for instructions.

**Comment by adisuissa** [comment]
@KBaichoo we initially added the default assignment operators to get around a similar issue.
I guess that if the copy c'tor isn't used then both can be removed.

**Comment by disconnect3d** [comment]
Fixed DCO and removed the default assignment operators

**Comment by ggreenway** [comment]
@yanavlasov please review

**Comment by disconnect3d** [comment]
@yanavlasov ping :)

**Comment by adisuissa** [comment]
I'm not sure if these changes are needed at the moment.
@disconnect3d WDYT?

**Comment by disconnect3d** [comment]
@adisuissa We suggest those changes to fix the oss-fuzz build of Envoy which seems to be broken for almost a year. Or at least the https://oss-fuzz-build-logs.storage.googleapis.com/index.html#envoy page reports that the last successful build was on 31/01/2021 and all the builds that happen now fail due to the issues that we try to fix in this PR.

The initial set of changes sent were enough to fix that build as we tested it internally in Trail of Bits.

**Comment by adisuissa** [comment]
To the best of my knowledge, they were broken before, and then we added #18865 and #18896 to address some of the issues, and also removed AFL (https://github.com/google/oss-fuzz/pull/6891) to get the build working.

The build now partially works (the fuzzers are built, but the validation step is still failing as can be seen in the recent logs).

**Comment by yanavlasov** [comment]
From the logs of the most recent build it looks like compilation is now successful, but then there are ASAN failures in the gtest framework.

**Comment by adisuissa** [comment]
> From the logs of the most recent build it looks like compilation is now successful, but then there are ASAN failures in the gtest framework.

Yes, this is due to part of the code compiled with ASAN and part of it without, see: https://github.com/google/oss-fuzz/blob/9d0a2b953a9ae4e8c6a853dcbd7053081b12a322/projects/envoy/build.sh#L83.

After solving this, there is another issue (timeout during the check_build step), that I'm trying to solve at the moment.

**Comment by disconnect3d** [comment]
> After solving this, there is another issue (timeout during the check_build step), that I'm trying to solve at the moment.

We ran things on our private server so we didn't get any timeouts.

Anyway, it seems that Envoy oss-fuzz builds are up again (and we are now getting crash reports), so I am closing this 

image
(screenshot via https://oss-fuzz-build-logs.storage.googleapis.com/index.html#envoy)

**Review by jmarantz — APPROVED** [review]

**Comment by adisuissa** [comment]
Not sure if this PR was supposed to be merged or closed.
I'll take a look at the fuzz-build results tomorrow and if it fails, we'll need to revert this PR.

**Pull request merged**